distributed/diagnostics/nvml.py

```diff
@@ -74,10 +82,17 @@ def has_cuda_context():
     init_once()
     for index in range(device_get_count()):
         handle = pynvml.nvmlDeviceGetHandleByIndex(index)
-        if hasattr(pynvml, "nvmlDeviceGetComputeRunningProcesses_v2"):
-            running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses_v2(handle)
-        else:
-            running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
+        try:
+            if hasattr(pynvml, "nvmlDeviceGetComputeRunningProcesses_v2"):
+                running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses_v2(
+                    handle
+                )
+            else:
+                running_processes = pynvml.nvmlDeviceGetComputeRunningProcesses(handle)
+        except pynvml.NVMLError_Unknown:
+            if _in_wsl():
+                return False
```

Comment:
Once we get to this point, should we do something to either disable NVML monitoring or inform the user of its need to be disabled?

Reply:
I don't think the issues are necessarily related, it seems to me that the issue with monitoring is a bug, whereas this one is a known missing feature. Therefore, I don't think it makes sense to do either of those things at this place exactly. However, we could add some warning if we identify WSL is being used so that the user is aware of the bug and let know that it can be disabled via the environment variable, or alternatively we could disable NVML altogether when the system is using WSL and in this case this PR would possibly be unnecessary.

Reply:
Hmm that actually does seem like a more elegant solution here - just adding `_in_wsl()` to this check:

https://github.com/dask/distributed/blob/d0b40d3b9c36e4ebcd61f95baab94c6344e17f1d/distributed/diagnostics/nvml.py#L18-L20

And adding the same general check for NVML initialization that we use for `device_get_count`:

https://github.com/dask/distributed/blob/d0b40d3b9c36e4ebcd61f95baab94c6344e17f1d/distributed/diagnostics/nvml.py#L38-L40

Is sufficient to resolve all WSL issues, including those noted in https://github.com/dask/distributed/issues/5342 since we no longer need to worry about monitoring NVML metrics for non-GPU clusters. The only downside I could imagine with this approach is if there were users who wanted to view the GPU metrics of workers when using a `LocalCluster`, though those metrics wouldn't really be indicative of anything in the computations.

Reply:
But this is disabling WSL completely, until we can verify all the metrics are available and eventually undo changes here, right? Meaning, if some metric was usable before, it can't be visualized at all anymore.

Besides what you already pointed out, another downside I see with this is if we eventually decide to reenable NVML diagnostics in WSL, we may still end up needing to add the check for `has_cuda_context` again, as we don't know whether all limitations will be resolved at the same time. I don't have strong opinions either way, just trying to point out potential effects we may experience in the future.

Reply:
> Meaning, if some metric was usable before, it can't be visualized at all anymore.

Correct; I originally assumed that this wouldn't be much of an issue considering that `LocalCUDAClusters` can't monitor these metrics anyways on WSL, and the metrics aren't really relevant for a standard `LocalCluster` (pretty sure all the workers just monitor the first sequential GPU?), but I haven't checked to see if it's possible to do NVML monitoring on other cluster types using `CUDAWorkers`.

> we may still end up needing to add the check for has_cuda_context again, as we don't know whether all limitations will be resolved at the same time.

That's a good point - if it makes sense to, I can add back the try/except for `has_cuda_context`, maybe with a note that the except block shouldn't be getting reached at this point, but might be relevant later.

Reply:
> Correct; I originally assumed that this wouldn't be much of an issue considering that LocalCUDAClusters can't monitor these metrics anyways on WSL, and the metrics aren't really relevant for a standard LocalCluster (pretty sure all the workers just monitor the first sequential GPU?), but I haven't checked to see if it's possible to do NVML monitoring on other cluster types using CUDAWorkers.

Why can't `LocalCUDACluster`s monitor those metrics on WSL? Maybe I'm missing something, but I thought it was just a few of them and only due to the known bug.

> That's a good point - if it makes sense to, I can add back the try/except for has_cuda_context, maybe with a note that the except block shouldn't be getting reached right now, but might be relevant later.

I don't know if that's necessary, it may be useful to prevent us from potentially debugging the same issue again in the future? I'm not certain what would be the correct course of action here, so I'll leave up to you to decide about this one.

Reply:
> Why can't `LocalCUDAClusters` monitor those metrics on WSL?

The error in https://github.com/dask/distributed/issues/5567#issuecomment-988096144 seemed to imply that after unblocking the issue with checking for CUDA context, we are unable to get device handles by index, and would need to set `DASK_DISTRIBUTED__DIAGNOSTICS__NVML=False` in order to continue using the CUDA cluster without getting constant errors from the attempts to monitor NVML metrics.

> I'm not certain what would be the correct course of action here, so I'll leave up to you to decide about this one.

I think leaving out the try/except block, but including a comment referencing the original issue should be sufficient.

Reply:
> The error in #5567 (comment) seemed to imply that after unblocking the issue with checking for CUDA context, we are unable to get device handles by index, and would need to set DASK_DISTRIBUTED__DIAGNOSTICS__NVML=False in order to continue using the CUDA cluster without getting constant errors from the attempts to monitor NVML metrics.

Is that _always_ the case? I thought NVML diagnostics in WSL worked in some cases, am I mistaken? If I am, then I agree it's sufficient.

> I think leaving out the try/except block, but including a comment referencing the original issue should be sufficient.

Agreed, a comment would be good and sufficient for now.

Reply:
> Is that _always_ the case?

Not always; to sum up the cases I know of where you can view real-time NVML metrics on WSL:

- using a `LocalCluster` without this PR; all the workers will be monitoring the first sequential GPU, which shouldn't be impacted since computation is happening on CPU
- using a `LocalCUDACluster` without this PR; immediately after construction, the cluster seems to monitor the relevant GPUs without issue. If you try to do anything with CUDA afterwards (i.e. construct a cuDF / dask-cuDF dataframe), the NVML monitoring will being failing

Given those cases, I feel like the choice to disable monitoring is reasonable in that:

- for a `LocalCluster`, these metrics generally don't have any relevance to the computations on CPU
- for a `LocalCUDACluster`, we generally can't monitor metrics once we do any computations on GPU, which in a sense defeats the purpose of having the metrics in the first place

That being said, I haven't tried any other set ups; I hypothesize that the only way we would be able to have working GPU workers without running into this issue is if we created them in a way that avoids a call to `has_cuda_context`, though I'd need to take some to think about if this is possible.

Reply:
Thanks for all details, Charles. Given all the issues you pointed out, I agree it's reasonable to just disable it entirely instead of trying to patch around every single minor issue for something that ultimately has to be fixed by NVML.

With that said, I think we're good here. Thanks for the patience to summarize everything!